### PR TITLE
Adding utitly functions and macro to catch cusparse errors and implementing transpose mode, see #614 and #650

### PR DESCRIPTION
**Pull request by lucbv**
This is mainly copies things done in Kokkos for CUDA_SAFE_CALL.
Also adding a small test to make sure it works when cusparse is enabled

### CODE REVIEW

File: src/common/KokkosKernels_SparseUtils_cusparse.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//               KokkosKernels 0.9: Linear Algebra and Graph Kernels
```
Comment (ndellingwood) [inline src/common/KokkosKernels_SparseUtils_cusparse.hpp]: @lucbv the copyright update PR just went in, so you'll need to update the copyright header in this file, just pattern match another file

File: src/common/KokkosKernels_SparseUtils_cusparse.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
```
Comment (ndellingwood) [inline src/common/KokkosKernels_SparseUtils_cusparse.hpp]: This line will also need to be updated

File: src/common/KokkosKernels_SparseUtils_cusparse.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
```
Comment (ndellingwood) [inline src/common/KokkosKernels_SparseUtils_cusparse.hpp]: And this line (SANDIA CORPORATION to NTESS)

**Comment by lucbv** [comment]
@ndellingwood 
I took care of the copyright update and I added some logic to make sure my test is only added when cusparse is enabled.

**Comment by lucbv** [comment]
@ndellingwood the problem is actually a bit bigger than that, we are not consistently adding the copyright notice in our tests, for instance look at the following files:
- `Test_Sparse_spadd.hpp`
- `Test_sparse_spmv.hpp`
- `Test_sparse_trsv.hpp`
- `Test_Sparse_Utils_cusparse.hpp`
that's only looking at the `unit_test/sparse folder`, there could be more in other folders...
If you want me to I can add the copyright notice to these files as part of this PR.

**Comment by lucbv** [comment]
Referencing issue #614 as it lists some of the work to be done here.
At the moment mainly creating `CUSPARSE_SAFE_CALL` macro and using it in conjunction with `CUDA_SAFE_CALL` from kokkos to clean-up the spmv cusparse interface.
Also switching from c cast to reinterpret_cast in a few places.

**Comment by mhoemmen** [comment]
@lucbv please PLEASE namespace your macros -- e.g., `KOKKOSKERNELS_CUSPARSE_SAFE_CALL` -- I know it's unwieldy but you can imagine every application will define `CUDA_SAFE_CALL` etc.

File: src/common/KokkosKernels_SparseUtils_cusparse.hpp [hunk 4]
```diff
@@ -73,11 +73,16 @@ inline void cusparse_internal_safe_call(cusparseStatus_t cusparseStatus,
   }
 }
 
+  // The macro below defines is the public interface for the safe cusparse calls.
+  // The functions themselves are protected by impl namespace.
+#define KOKKOS_CUSPARSE_SAFE_CALL(call) \
```
Comment (mhoemmen) [inline src/common/KokkosKernels_SparseUtils_cusparse.hpp]: Thank you for namespacing the macro : - )

Comment (lucbv) [reply src/common/KokkosKernels_SparseUtils_cusparse.hpp]: No problem, this should allow you and others to use it if needed in other packages that rely on kk

**Comment by lucbv** [comment]
@ndellingwood 
I am running the spot-check and spot-check-tpls for this PR on kokkos-dev.
Until the OHPC storage space issue is resolved, this is all I can test, can you run a spot-check on kokkos-dev2? I do not have access to that machine...

**Comment by ndellingwood** [comment]
> Until the OHPC storage space issue is resolved,

Yeah, this is killing a bunch of the nightlies as well



> can you run a spot-check on kokkos-dev2

yes, launching it now

**Comment by ndellingwood** [comment]
kokkos-dev-2 tpl spot-check results:

```
Running on machine: kokkos-dev-2
Repository Status:  0470b438f649a4a6541d954371166936e3a59df6 Adding support for mode[] in cusparse spmv interface. Also renamince CUSPARSE_SAFE_CALL to avoid macro clashes. Finally cleaning up code a bit more in spmv_tpl_spec_decl


Going to test compilers:  gcc/7.3.0 gcc/8.3.0 gcc/9.1 intel/18.0.5 clang/8.0 cuda/10.1
Testing compiler gcc/7.3.0
  Starting job gcc-7.3.0-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.3.0-OpenMP-release
Testing compiler gcc/8.3.0
  Starting job gcc-7.3.0-Pthread-release
kokkos options: 
kokkos devices: Pthread
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.3.0-Pthread-release
Testing compiler gcc/9.1
  Starting job gcc-8.3.0-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-8.3.0-Serial-release
  Starting job gcc-9.1-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-9.1-OpenMP-release
Testing compiler intel/18.0.5
  Starting job gcc-9.1-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-9.1-Serial-release
Testing compiler clang/8.0
  Starting job intel-18.0.5-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED intel-18.0.5-OpenMP-release
  Starting job clang-8.0-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed 
  PASSED clang-8.0-Cuda_OpenMP-release
Testing compiler cuda/10.1
  Starting job clang-8.0-Pthread_Serial-release
kokkos options: 
kokkos devices: Pthread,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed 
  PASSED clang-8.0-Pthread_Serial-release
  Starting job cuda-10.1-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=121 run_time=154
clang-8.0-Pthread_Serial-release build_time=51 run_time=155
cuda-10.1-Cuda_OpenMP-release build_time=157 run_time=134
gcc-7.3.0-OpenMP-release build_time=45 run_time=50
gcc-7.3.0-Pthread-release build_time=41 run_time=78
gcc-8.3.0-Serial-release build_time=42 run_time=77
gcc-9.1-OpenMP-release build_time=52 run_time=50
gcc-9.1-Serial-release build_time=46 run_time=73
intel-18.0.5-OpenMP-release build_time=150 run_time=33
```

**Comment by lucbv** [comment]
### kokkos-dev --spot-check
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-hwloc-release build_time=127 run_time=163
clang-4.0.1-Pthread_Serial-release build_time=126 run_time=205
clang-7.0.1-Cuda_OpenMP-hwloc-release build_time=341 run_time=290
clang-7.0.1-Cuda_OpenMP-release build_time=344 run_time=293
cuda-9.2-Cuda_OpenMP-release build_time=389 run_time=408
gcc-5.3.0-OpenMP-hwloc-release build_time=98 run_time=56
gcc-5.3.0-OpenMP-release build_time=101 run_time=56
gcc-7.3.0-Serial-hwloc-release build_time=82 run_time=95
gcc-7.3.0-Serial-release build_time=82 run_time=95
intel-17.0.1-OpenMP-hwloc-release build_time=172 run_time=57
intel-17.0.1-OpenMP-release build_time=167 run_time=57

**Comment by lucbv** [comment]
Thanks @ndellingwood, it will speed things up for this PR, I am now testing with `--spot-check-tpls` hopefully this is all clean and I can close a couple issues in the 3.1 kanban : )

**Comment by lucbv** [comment]
@ndellingwood 
I am actually seeing a build failure on `kokkos-dev` with `--spot-check-tpls`, the old interface does not support the following functions: `cusparseGetErrorName` and `cusparseGetErrorString`.
I will implement an extra function that returns a different string based on the enum returned by cusparse. It should not take long to do and retest.
Good thing we have the new `tpls` option in `cm_test_all_sandia` ; )

**Comment by lucbv** [comment]
Restarting the spot-check-tpl on Kokkos-dev with special handling of errors for old cusparse interface. Let's see how it does tomorrow morning.

**Comment by mhoemmen** [comment]
@lucbv Note that the error macros also change from CUDA 9.2 to the new version.  Please feel free to copy and paste from my PR (it's probably in `Tpetra_Details_CuSparseMatrix.cpp`).

**Comment by lucbv** [comment]
Alright, latests patch seems to have fixed the previous issue, spot-check passes on kokkos-dev:

### kokkos-dev --spot-check
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-hwloc-release build_time=125 run_time=165
clang-4.0.1-Pthread_Serial-release build_time=126 run_time=196
clang-7.0.1-Cuda_OpenMP-hwloc-release build_time=348 run_time=292
clang-7.0.1-Cuda_OpenMP-release build_time=355 run_time=317
cuda-9.2-Cuda_OpenMP-release build_time=395 run_time=409
gcc-5.3.0-OpenMP-hwloc-release build_time=97 run_time=57
gcc-5.3.0-OpenMP-release build_time=96 run_time=81
gcc-7.3.0-Serial-hwloc-release build_time=83 run_time=92
gcc-7.3.0-Serial-release build_time=83 run_time=89
intel-17.0.1-OpenMP-hwloc-release build_time=168 run_time=56
intel-17.0.1-OpenMP-release build_time=171 run_time=56

### kokkos-dev --spot-check-tpls
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-hwloc-release build_time=126 run_time=273
clang-4.0.1-Pthread_Serial-release build_time=126 run_time=190
clang-7.0.1-Cuda_OpenMP-hwloc-release build_time=405 run_time=282
clang-7.0.1-Cuda_OpenMP-release build_time=347 run_time=278
cuda-9.2-Cuda_OpenMP-release build_time=407 run_time=317
gcc-5.3.0-OpenMP-hwloc-release build_time=97 run_time=48
gcc-5.3.0-OpenMP-release build_time=100 run_time=46
gcc-7.3.0-Serial-hwloc-release build_time=84 run_time=82
gcc-7.3.0-Serial-release build_time=82 run_time=83
intel-17.0.1-OpenMP-hwloc-release build_time=169 run_time=32
intel-17.0.1-OpenMP-release build_time=169 run_time=33

**Comment by lucbv** [comment]
@ndellingwood if you would not mind doing one more check on kokkos-dev2, I think we should be good after that.

**Comment by ndellingwood** [comment]
Running tpl spot check now

**Comment by ndellingwood** [comment]
Passes kokkos-dev-2 spot-check-tpls:

```
Running on machine: kokkos-dev-2
Repository Status:  d4da4bf9a2975e990df9df12e3fab3ba75195087 Adding guard in error reporting utility.


Going to test compilers:  gcc/7.3.0 gcc/8.3.0 gcc/9.1 intel/18.0.5 clang/8.0 cuda/10.1
Testing compiler gcc/7.3.0
  Starting job gcc-7.3.0-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.3.0-OpenMP-release
Testing compiler gcc/8.3.0
  Starting job gcc-7.3.0-Pthread-release
kokkos options: 
kokkos devices: Pthread
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.3.0-Pthread-release
Testing compiler gcc/9.1
  Starting job gcc-8.3.0-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-8.3.0-Serial-release
  Starting job gcc-9.1-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-9.1-OpenMP-release
Testing compiler intel/18.0.5
  Starting job gcc-9.1-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-9.1-Serial-release
Testing compiler clang/8.0
  Starting job intel-18.0.5-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED intel-18.0.5-OpenMP-release
  Starting job clang-8.0-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed 
  PASSED clang-8.0-Cuda_OpenMP-release
Testing compiler cuda/10.1
  Starting job clang-8.0-Pthread_Serial-release
kokkos options: 
kokkos devices: Pthread,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed 
  PASSED clang-8.0-Pthread_Serial-release
  Starting job cuda-10.1-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=121 run_time=155
clang-8.0-Pthread_Serial-release build_time=51 run_time=154
cuda-10.1-Cuda_OpenMP-release build_time=157 run_time=134
gcc-7.3.0-OpenMP-release build_time=45 run_time=49
gcc-7.3.0-Pthread-release build_time=42 run_time=80
gcc-8.3.0-Serial-release build_time=42 run_time=74
gcc-9.1-OpenMP-release build_time=52 run_time=49
gcc-9.1-Serial-release build_time=46 run_time=74
intel-18.0.5-OpenMP-release build_time=149 run_time=34
```

@lucbv ready for merge?

**Comment by lucbv** [comment]
Thanks @ndellingwood it is indeed ready to merge : )

**Pull request merged**